plugins/Referrers/RecordBuilders/Referrers.php

```diff
@@ -0,0 +1,256 @@
+ *
+ */
+
+namespace Piwik\Plugins\Referrers\RecordBuilders;
+
+use Piwik\ArchiveProcessor;
+use Piwik\ArchiveProcessor\Record;
+use Piwik\ArchiveProcessor\RecordBuilder;
+use Piwik\Common;
+use Piwik\Config;
+use Piwik\DataAccess\LogAggregator;
+use Piwik\DataTable;
+use Piwik\Metrics;
+use Piwik\Plugins\Referrers\API;
+use Piwik\Plugins\Referrers\Archiver;
+
+class Referrers extends RecordBuilder
+{
+    public function __construct()
+    {
+        parent::__construct();
+
+        $this->columnToSortByBeforeTruncation = Metrics::INDEX_NB_VISITS;
+
+        // Reading pre 2.0 config file settings
+        $this->maxRowsInTable = @Config::getInstance()->General['datatable_archiving_maximum_rows_referers'];
+        $this->maxRowsInSubtable = @Config::getInstance()->General['datatable_archiving_maximum_rows_subtable_referers'];
+        if (empty($this->maxRowsInTable)) {
+            $this->maxRowsInTable = Config::getInstance()->General['datatable_archiving_maximum_rows_referrers'];
+            $this->maxRowsInSubtable = Config::getInstance()->General['datatable_archiving_maximum_rows_subtable_referrers'];
+        }
+    }
+
+    public function getRecordMetadata(ArchiveProcessor $archiveProcessor): array
+    {
+        return [
+            Record::make(Record::TYPE_BLOB, Archiver::SEARCH_ENGINES_RECORD_NAME),
+            Record::make(Record::TYPE_BLOB, Archiver::SOCIAL_NETWORKS_RECORD_NAME),
+            Record::make(Record::TYPE_BLOB, Archiver::KEYWORDS_RECORD_NAME),
+            Record::make(Record::TYPE_BLOB, Archiver::CAMPAIGNS_RECORD_NAME),
+            Record::make(Record::TYPE_BLOB, Archiver::WEBSITES_RECORD_NAME),
+            Record::make(Record::TYPE_BLOB, Archiver::REFERRER_TYPE_RECORD_NAME),
+
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_SEARCH_ENGINE_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::SEARCH_ENGINES_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_SOCIAL_NETWORK_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::SOCIAL_NETWORKS_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_KEYWORD_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::KEYWORDS_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_CAMPAIGN_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::CAMPAIGNS_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_WEBSITE_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::WEBSITES_RECORD_NAME),
+            Record::make(Record::TYPE_NUMERIC, Archiver::METRIC_DISTINCT_URLS_RECORD_NAME)
+                ->setIsCountOfBlobRecordRows(Archiver::WEBSITES_RECORD_NAME, true)
+                ->setMultiplePeriodTransform(function (float $value, array $counts) {
+                    return $counts['recursive'] - $counts['level0'];
```

Comment:
Might be good to add a comment here, so is easier to understand what exactly this is good for.